File: peekingIterator.py
#lt solution with presaved _next element.
#the reason why we need self._next is that there is a self.next() func
#hence the iterator can be just self.iterator
class PeekingIterator:
    def __init__(self, iterator):
        self._next = iterator.next()
        self._iterator = iterator

    def peek(self):
        return self._next

    def next(self):
        if self._next is None:
            raise StopIteration()
        toReturn = self._next
        self._next = None
        if self._iterator.hasNext():
            self._next = self._iterator.next()
        return toReturn

    def hasNext(self):
        return self._next is not None

File: peekingIterator.kt
// Kotlin Iterator reference:
// https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-iterator/
// I learned this solution the hard way...
class PeekingIterator(iteratorInput:Iterator<Int>):Iterator<Int> {
    val iterator = iteratorInput
    var presave: Int? = null
    fun peek(): Int {
    	if(presave == null) {
            presave = iterator.next()
            return presave!!
        } else {
            return presave!!
        }
    }
    
    override fun next(): Int {
        if(presave != null) {
            val temp = presave!!
            presave = null
            return temp
        }
        else return iterator.next()
    }
    
    override fun hasNext(): Boolean {
        return presave != null || iterator.hasNext()
    }
}

/**
 * Your PeekingIterator object will be instantiated and called as such:
 * var obj = PeekingIterator(arr)
 * var param_1 = obj.next()
 * var param_2 = obj.peek()
 * var param_3 = obj.hasNext()
 */
